src/driver/mod.rs

```diff
@@ -144,40 +144,73 @@ impl Drop for Driver {
+                lc @ Lifecycle::Completed(_) => {
+                    // don't cancel completed items
+                    *cycle = lc;
+                }
+
+                Lifecycle::CompletionList(indices) => {
+                    let mut list = indices.clone().into_list(&mut self.ops.completions);
+                    if !io_uring::cqueue::more(list.peek_end().unwrap().flags) {
+                        // This op is complete. Replace with a null Completed entry
+                        *cycle = Lifecycle::Completed(op::CqeResult {
+                            result: Ok(0),
+                            flags: 0,
+                        });
+                    }
+                }
+
+                _ => {
+                    // All other states need cancelling.
+                }
             }
         }
 
-        // cancel all ops
-        for id in cancellable_ops {
-            unsafe {
-                while self
-                    .uring
-                    .submission()
-                    .push(&AsyncCancel::new(id as u64).build().user_data(u64::MAX))
-                    .is_err()
-                {
-                    self.submit().expect("Internal error when dropping driver");
+        // Submit cancellation for all ops marked Ignored
+        for (id, cycle) in self.ops.lifecycle.iter_mut() {
+            if let Lifecycle::Ignored(..) = cycle {
+                unsafe {
+                    while self
+                        .uring
+                        .submission()
+                        .push(&AsyncCancel::new(id as u64).build().user_data(u64::MAX))
+                        .is_err()
+                    {
+                        self.uring
+                            .submit_and_wait(1)
+                            .expect("Internal error when dropping driver");
+                    }
                 }
             }
         }
 
-        // TODO: add a way to know if a multishot op is done sending completions
-        // SAFETY: this is currently unsound for multishot ops
-        while !self
-            .ops
-            .lifecycle
-            .iter()
-            .all(|(_, cycle)| matches!(cycle, Lifecycle::Completed(_)))
-        {
-            // If waiting fails, ignore the error. The wait will be attempted
-            // again on the next loop.
-            let _ = self.wait();
-            self.tick();
+        // Wait until all in flight cancellations have stopped
+        loop {
+            if self.ops.lifecycle.is_empty() {
+                break;
+            }
+            // Cycles are either all ignored or complete
+            let remove = if let (id, Lifecycle::Completed(..)) =
+                &self.ops.lifecycle.iter().next().unwrap()
+            {
+                Some(*id)
+            } else {
+                None
+            };
+
+            // Remove completed lifecycles from Slab
+            // This prevents worst case quadratic processing
```

Comment:
I like the intention but I think it is still quadratic because the lifecycle.iter() will run through its Vec until finding a Some, always starting from the beginning.

And if the first entry found by the lifecycle.iter().next() is always an ignored, we have to do the search again to find its location again, after we've called wait and tick once below.

When things work normally, this will be fine. I like to consider the if the slab had grown to 10_000 simultaneous operations. I don't know if that is small or considered too large by tokio's standards for a single threaded runtime.

But when there is a problem, like I put into the other PR, I wish there was logic to trigger at least a warning to stderr that there appears to be no progress being made in getting something canceled. But Noah didn't think it that important at the time so it's not more important now either.

Reply:
Ok: I have worked out a scheme for single traversal. Will implement